### PR TITLE
[2.0] New content for templating: Create custom templates for dotnet new

### LINKED ISSUE #2405: [2.0] New content for templating: Create custom templates for dotnet new
Title: *Create custom templates for dotnet new*
Area: *docs/core/tools*

Reference
* [How to create your own templates for dotnet new](https://blogs.msdn.microsoft.com/dotnet/2017/04/02/how-to-create-your-own-templates-for-dotnet-new/)
* [dotnet/templating Wiki](https://github.com/dotnet/templating/wiki)

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #2405

### CODE REVIEW

**Comment by GuardRex** [comment]
@mairaw Ready for :eyes:.

Note that this one concentrates on getting the custom template topic setup and only adds the `-i|--install` and `-u|--uninstall` options to the `dotnet new` topic. There will be more (broader) `dotnet new` updates on https://github.com/dotnet/docs/pull/2480 as soon as these other update PR's get merged.

**Comment by GuardRex** [comment]
One more thing I noticed: The metadata title has the preview warning, and the WARNING note is at the top of the content. However, the H1 doesn't have it. I'll fix that on review when I'm working on feedback.

File: docs/core/tools/custom-templates.md
Comment (mairaw) [inline docs/core/tools/custom-templates.md]: i'd add the verb create to the file name too

File: docs/core/tools/custom-templates.md
Comment (mairaw) [inline docs/core/tools/custom-templates.md]: add description

File: docs/core/tools/custom-templates.md
Comment (mairaw) [inline docs/core/tools/custom-templates.md]: add this topic to the TOC

File: docs/core/tools/custom-templates.md
Comment (mairaw) [inline docs/core/tools/custom-templates.md]: wouldn't be better to say that The .NET Core SDK comes with many templates pre-installed for you to use with dotnet new?

File: docs/core/tools/custom-templates.md
Comment (mairaw) [inline docs/core/tools/custom-templates.md]: Microsoft framework is throwing me off a bit here. Maybe we say .NET platform?

File: docs/core/tools/custom-templates.md
Comment (mairaw) [inline docs/core/tools/custom-templates.md]: wouldn't go into details of opening issues here. Also, we don't typically call out blog post authors in the docs.

File: docs/core/tools/custom-templates.md
Comment (mairaw) [inline docs/core/tools/custom-templates.md]: are composed? maybe includes?

and what if i want a non-runnable project as a source code like a config file or solution?

File: docs/core/tools/custom-templates.md
Comment (mairaw) [inline docs/core/tools/custom-templates.md]: provides configuration -> provides configuration information

when creating output from the source files and folders -> is this necessary?

**Review by mairaw — COMMENTED** [review]
some initial comments. need to finish the review

File: docs/core/tools/custom-templates.md
Comment (GuardRex) [inline docs/core/tools/custom-templates.md]: I'm not 100% feel'in this one because I really did mean to say that the template engine (as a tool) works with any "Microsoft framework" (as opposed to any other non-MS frameworks) and list (at least some of) them. However, I don't think this sentence helps much here. The main thrust of this paragraph has more to do with using them, and I think we can assume that everyone knows this is a MS technology. Let's try killing this sentence and see if that works better.

>

File: docs/core/tools/custom-templates.md
Comment (GuardRex) [inline docs/core/tools/custom-templates.md]: > what if i want a non-runnable project as a source code like a config file or solution

Described below; however, I'm going to flesh it out a bit and explicitly call out "solution file" as a possible source/output for a template.

**Comment by GuardRex** [comment]
@mairaw Ready for another :eyes: look.

File: docs/core/tools/create-custom-templates.md
Comment (svick) [inline docs/core/tools/create-custom-templates.md]: > A detailed pre-release blog post that describes how to create and install custom templates is at How to create your own templates for dotnet new.

What makes the blog post useful to me if I'm already reading this docs article about how to create templates? Maybe you should mention why I should read the blog post (instead of continuing reading this article), or remove the link altogether?

File: docs/core/tools/create-custom-templates.md
Comment (svick) [inline docs/core/tools/create-custom-templates.md]: Does this mean I can't pack a custom template on Unix? Is that worth mentioning?

File: docs/core/tools/create-custom-templates.md
Comment (svick) [inline docs/core/tools/create-custom-templates.md]: Shouldn't it be "The contents […] are placed"?

File: docs/core/tools/create-custom-templates.md
Comment (svick) [inline docs/core/tools/create-custom-templates.md]: Is there a reason to use absolute path here? Wouldn't just `nuget pack GarciaSoftware.ConsoleTemplate.CSharp.nuspec` be better? Or maybe use a shorter, relative path, like `nuget pack Templates/GarciaSoftware.ConsoleTemplate.CSharp/GarciaSoftware.ConsoleTemplate.CSharp.nuspec`, which can be also used in later examples?

File: docs/core/tools/create-custom-templates.md
Comment (svick) [inline docs/core/tools/create-custom-templates.md]: This example uses slashes and explicitly names the user (`C:/Users/Garcia`). The example above uses backslahes and user name placeholder (`C:\Users\<USER>`). If you're going to keep these absolute paths, I think they should at least be consistent.

File: docs/core/tools/create-custom-templates.md
Comment (GuardRex) [inline docs/core/tools/create-custom-templates.md]: Yes, you're correct: We decided in a short offline chat (I'll add a note to the issue) that we need to keep the blog post linked here for two reasons:

* For the first pass on fully covering templates, we're covering the utter basics of template creation and usage and not adding a tutorial/walk-through to go with this reference topic. I did one thing that helps a dev trying to create a basic template: I kept the example "Garcia Software Console" template in use throughout the topic. This topic is really only going to get us started. We need a real tutorial, and I'll open an issue after we get this merged to discuss this aspect further.
* The first version of this topic won't cover all of the features of template design, so we'll need to expand the coverage. Where and how to cover the finer details for features such as replaceable parameter values, optional content, parameter options, custom processing, etc. is TBD. When I open the issue later, let's discuss how to best layout the coverage.

We'll probably end up with two issues: one saying we need a tutorial and a second saying we need to cover additional features of template design and how to lay that topic (or topics) out.

What I'll do here is just instruct the reader to access the blog post and Wiki for additional info after they've read this topic. (Let's take a 1st look at some draft language on the next commit, and then we can modify as needed.) As the reader gets down into the topic, they have linked access to  the *template.json* schema. When the features of templates are fully in place in the docs, then yes, I think we should drop the blog post entirely.

File: docs/core/tools/create-custom-templates.md
Comment (GuardRex) [inline docs/core/tools/create-custom-templates.md]: Good question. I just looked at the issues for Mono: https://github.com/NuGet/Home/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+mono

... so ... I guess as long as none of those would result in a big :boom: it should work. Before we state anything, let's ping @sayedihashimi to see if he knows.

File: docs/core/tools/create-custom-templates.md
Comment (GuardRex) [inline docs/core/tools/create-custom-templates.md]: Not knowing where they are, I feel a bit safer with a full path. This is part of that "weak tutorial" idea that I had for this topic, where you see I've placed examples to show how a basic template works. Since it isn't a tutorial, I'm less concerned about their working directory when they try to run the command. Of course, they must change it a little to make it work.

File: docs/core/tools/create-custom-templates.md
Comment (sayedihashimi) [inline docs/core/tools/create-custom-templates.md]: >for any type of .NET project

Technically speaking we are not limited to creating .NET Projects. It can create templates for any project type.

Comment (sayedihashimi) [inline docs/core/tools/create-custom-templates.md]: I just read a comment below where this is explained.

File: docs/core/tools/create-custom-templates.md
Comment (sayedihashimi) [inline docs/core/tools/create-custom-templates.md]: It would be nice to add

```
"$schema": "http://json.schemastore.org/template",
```

With this included any editor that support JSON Schema will be enabled. For VS Code this is needed to get IntelliSense when editing the `template.json` file.

File: docs/core/tools/create-custom-templates.md
Comment (GuardRex) [inline docs/core/tools/create-custom-templates.md]: @sayedihashimi Should I make it say "templates for any type of project"? (i.e., drop the ".NET") there?

File: docs/core/tools/create-custom-templates.md
Comment (sayedihashimi) [inline docs/core/tools/create-custom-templates.md]: For xplat I'm hearing that NuGetizer is the way to go, see https://github.com/dotnet/dotnet-template-samples/issues/16#issuecomment-310786056

File: docs/core/tools/create-custom-templates.md
Comment (sayedihashimi) [inline docs/core/tools/create-custom-templates.md]: Can we add this sample to https://github.com/dotnet/dotnet-template-samples as well?

File: docs/core/tools/dotnet-new.md
Comment (sayedihashimi) [inline docs/core/tools/dotnet-new.md]: Please add a link to the samples repo at https://github.com/dotnet/dotnet-template-samples/

File: docs/core/tools/create-custom-templates.md
Comment (GuardRex) [inline docs/core/tools/create-custom-templates.md]: Sure ... I'll set up a PR for it over there.

File: docs/core/tools/create-custom-templates.md
Comment (GuardRex) [inline docs/core/tools/create-custom-templates.md]: I'm removing ".NET" from the phrase.

**Comment by sayedihashimi** [comment]
Thanks for putting the doc together. I've reviewed the content and added a few comments.
Since this doc is aimed at the template developer I was thinking that the flow of this doc would be:

 1. Intro / overview
 1. How to create a template from an existing project
 1. How to create a new project from the template locally
    1. `dotnet new -i <path-to-folder>`
    1. `dotnet new <template-name>`
 1. How to create and publish the template as a NuGet package on nuget.org
 1. How to install and use your new template from nuget.org

In the current doc we mix info about folder/nuget package, and the typical template developer workflow is not clear.

File: docs/core/tools/create-custom-templates.md
Comment (sayedihashimi) [inline docs/core/tools/create-custom-templates.md]: OK thanks

File: docs/core/tools/create-custom-templates.md
Comment (sayedihashimi) [inline docs/core/tools/create-custom-templates.md]: Great thanks

**Comment by GuardRex** [comment]
@sayedihashimi When you say "How to create a template from an existing project," I covered it implicitly, so I'll add a section to make it explicit after the intro.

RE the outline you provided: This wasn't written to be a dev tutorial/walk-through but a ref topic. I group the install methods and the uninstall methods for easier reference; otherwise, a dev has to wade through a lot of text to find relevant reference content. Here's the current outline: 

* Introduction/Overview
* Configuration
  * Source files and folders
  * Configuration file (*template.json* schema and info)
* .NET default templates
* *ADDING* Create a template from a project
* Packing a template into a NuGet package
* Installing a template
  * From a NuGet package at nuget.org
  * From a local *nupkg* file
  * From a file system directory
* Uninstalling a template
  * From a NuGet package at nuget.org
  * From a local *nupkg* file
  * From a file system directory
* Creating a project from a template

I agree from a developer walk-through/tutorial standpoint that this isn't ideal; however from a ref topic approach, I think this is smart: The headings here are what will show in the topic TOC, so it's easy to track down just 'packing', 'installing', 'uninstalling' or 'using' ... which is what those H1's represent below the ".NET default templates" section. I didn't intend this to be a tutorial, so that's why the outline turned out like this.

My suggestion is to create a separate dev tutorial/walk-through and have both topics. The dev tutorial then only needs to to refer to this as a pre-req then follow your, @sayedihashimi, outline step-by-step. It would be rather easy/quick to setup (and I do have time to create it ... they don't 🔪 chop me for two more weeks). 

@mairaw Thoughts?

**Comment by GuardRex** [comment]
@mhutch PTAL ... I mention/link the Wiki of https://github.com/mhutch/NuGetizer3000 in this topic dealing with `dotnet new` templates. Is this what you had in mind for an xplat NuGet packager that would be good for templates? Is that the best link text & link for this?

> Currently, a custom template is packed on Windows with \[nuget.exe](https://dist.nuget.org/win-x86-commandline/latest/nuget.exe) (not \[dotnet pack](dotnet-pack.md)). For cross-platform packaging, consider \[NuGetizer 3000](https://github.com/NuGet/Home/wiki/NuGetizer-3000).

**Comment by mlorbetske** [comment]
There's also this (https://github.com/tintoy/dotnet-template-templates/releases/tag/v0.1-beta1) the template template pack that has a packaging project template that uses dotnet pack instead of nuget.exe

File: docs/core/tools/create-custom-templates.md
Comment (GuardRex) [inline docs/core/tools/create-custom-templates.md]: https://github.com/dotnet/dotnet-template-samples/pull/20

**Comment by GuardRex** [comment]
One other things occurs, too, given the discussion on what this is ... is it a basic walk-thru/tutorial or is it a ref topic?

There was *A LOT* of content left out of this first version. All of those templating features that aren't discussed here (many of which are in your, @sayedihashimi, blog post): All of that content was supposed to find its way into this "official docs topic" over time (probably next revision).

This topic is (or "was" depending on what u cats want to do with it) supposed to be the place for that content, unless all of the features of templating couldn't be covered well in just one topic. In which case, this topic is the overview and those other sub-topics would fall in under this ... this all ends up in its own folder in the TOC ... etc.

Either way, this is setup "reference"-style for that reason, too, so let me know how u want to play it. If you want a walk-thru/tutorial, I can add that easily as a companion topic to this and we can cross-link them and link it to the `dotnet new` command.

**Comment by sayedihashimi** [comment]
>This wasn't written to be a dev tutorial/walk-through but a ref topic

OK sounds good. I think having the ref topic and walk through are both good things. Not sure how things are prioritized on the doc side though. Adding a companion topic later sounds like a good idea.

**Comment by GuardRex** [comment]
Cool. @mairaw will let me know if its :+1: to do it now.

**Comment by GuardRex** [comment]
@mairaw has said :tada: *"Yes!"* 🎈  to a companion tutorial. I'll use your, @sayedihashimi, outline and the sample that's shown here (`GarciaSoftware.ConsoleTemplate.CSharp`). She's going to let me know in a bit where the tutorial will go. I'll add it to this PR unless otherwise instructed.

**Comment by sayedihashimi** [comment]
@guardrex great news!

**Comment by mhutch** [comment]
@GuardRex all 3 options - nuget.exe, NuGetizer and `dotnet pack` work xplat if you're using Mono :)

I generally prefer NuGetizer. It isn't any better than the other options at this _specifically_ but there are some other pack scenarios it handles much better and i think it's good to have a single solution.

**Comment by GuardRex** [comment]
@mlorbetske @sayedihashimi I'm blocking this on https://github.com/dotnet/templating/issues/965 given your, @mlorbetske, and [@]richlander's comments. There's no milestone on that issue tho; so if that work will be addressed in a 2.x release, let me know and I'll unblock this.

One thing we want to cover is a dev tutorial/walk-through, so I've marked this **WIP** until that piece is added, which I'll get to this week.

Question: Should I add a bit on the installation of the 1.x template pack? I can add a couple of lines and the exact command under the `-i|--install <PATH|NUGET_ID>` option ... thoughts?

@seancpeters This is a good time for you to PTAL. I'll be adding a dev walk-thru topic to this PR, but that's a trivial piece. The guts of what we're doing are in the updates to the `dotnet new` command and the new topic on custom templates. Those bits are rolling along well here.

**Comment by mlorbetske** [comment]
@GuardRex for #965, we'll be adding a link to documentation describing installing templates when there's a parameter error (I believe for the 2.0 release yet), but that's about it. Is there something specific that you feel is blocking for this document referenced in there?

**Comment by GuardRex** [comment]
@mlorbetske I see. I couldn't tell if any major changes were expected in behavior or in help comments surrounding the template options, but you're saying 'no', so we're good. :+1:

The only question then is whether or not I should add a few lines and a command to show how to install the 1.x templates?

**Comment by mlorbetske** [comment]
@GuardRex I think the edit with my response to cover that was added as you were writing your reply :) I think the 1.x template install would be best covered separately, in a blog post about 1.x/2.0 SxS development (as it's not just the templates - multiple packs - that must be installed, but the runtime as well and the matter of different templates being available depending on the active CLI)

**Comment by GuardRex** [comment]
@mlorbetske Cool. Outside of reviewer feedback, I just need to get the tutorial/walk-thru in place. It's coming along very well. Thanks for all of your help.

**Comment by mairaw** [comment]
is this still WIP @GuardRex?

**Comment by GuardRex** [comment]
@mairaw Yes, very much so. I have to rip the tutorial/walk-thru bits out of what was going to be just the one topic to make two: a ref topic and a tutorial topic. I'll be finishing it up this weekend..

**Comment by mairaw** [comment]
I need to make more changes to the dotnet new command. But for now, I've just added the two syntaxes and readded old options so I don't lose them. In a separate PR, I'll make the additional changes needed.

**Review by mairaw — APPROVED** [review]
Thank you @guardrex for your work on this!

**Pull request merged**
